## CVD Prediction - Mendeley Dataset (Source: https://data.mendeley.com/datasets/dzz48mvjht/1)
Model Training and Evaluation

In [1]:
#load preprocessed data 
import pandas as pd
train_df = pd.read_csv("./data_subsets/train_50_50.csv")

X_test = pd.read_csv("./data_splits/X_test.csv")
y_test = pd.read_csv("./data_splits/y_test.csv")

#check out the data
train_df.head()

source_id  age  gender  chestpain  restingBP  serumcholestrol  \
0        151   20       1          1        170            352.0   
1        373   51       1          2        176            346.0   
2        625   60       0          0        131            164.0   
3        621   67       0          1        172            461.0   
4        469   74       0          2        127            420.0   

   fastingbloodsugar  restingrelectro  maxheartrate  exerciseangia  oldpeak  \
0                  1                0           138              0      1.4   
1                  0                2           160              1      2.0   
2                  0                0            86              1      2.3   
3                  0                1           134              0      0.8   
4                  0                2           113              1      2.7   

   slope  noofmajorvessels  target  
0      1                 0       1  
1      3                 3       1  
2      1                 2       0  
3      1                 1       0  
4      2                 1       1

In [2]:
TARGET = "target"
SENSITIVE = "gender"   # 1 = Male, 0 = Female

categorical_cols = ['gender','chestpain','fastingbloodsugar','restingrelectro','exerciseangia','slope','noofmajorvessels']
continuous_cols  = ['age','restingBP','serumcholestrol','maxheartrate','oldpeak']

X_train = train_df.drop(columns=[TARGET])
y_train = train_df[TARGET]

In [3]:
# SCALE NUMERIC FEATURES ONLY 

import pandas as pd
from sklearn.preprocessing import StandardScaler


# 1) fit scaler on TRAIN numeric columns only
scaler = StandardScaler()
X_train_num_scaled = pd.DataFrame(
    scaler.fit_transform(X_train[continuous_cols]),
    columns=continuous_cols,
    index=X_train.index
)

# 2) transform TEST with the same scaler
X_test_num_scaled = pd.DataFrame(
    scaler.transform(X_test[continuous_cols]),
    columns=continuous_cols,
    index=X_test.index
)

# 3) reassemble: raw categoricals + scaled numerics
X_train_scaled = pd.concat([X_train[categorical_cols].reset_index(drop=True),
                            X_train_num_scaled.reset_index(drop=True)], axis=1)
X_test_scaled  = pd.concat([X_test[categorical_cols].reset_index(drop=True),
                            X_test_num_scaled.reset_index(drop=True)], axis=1)

In [4]:
#onehot encode categorical, keep scaled numeric as is

import pandas as pd
from sklearn.preprocessing import OneHotEncoder

# 1) fit encoder on TRAIN categoricals only
ohe = OneHotEncoder(handle_unknown="ignore", drop="if_binary", sparse_output=False)
ohe.fit(X_train_scaled[categorical_cols])

# 2) transform TRAIN and TEST
X_train_cat = pd.DataFrame(
    ohe.transform(X_train_scaled[categorical_cols]),
    columns=ohe.get_feature_names_out(categorical_cols),
    index=X_train_scaled.index
)
X_test_cat = pd.DataFrame(
    ohe.transform(X_test_scaled[categorical_cols]),
    columns=ohe.get_feature_names_out(categorical_cols),
    index=X_test_scaled.index
)

# 3) concatenate: encoded categoricals + scaled numerics
X_train_ready = pd.concat([X_train_cat, X_train_scaled[continuous_cols]], axis=1)
X_test_ready  = pd.concat([X_test_cat,  X_test_scaled[continuous_cols]],  axis=1)

print("Final feature shapes:", X_train_ready.shape, X_test_ready.shape)

Final feature shapes: (600, 22) (200, 22)


### Traditional ML Models - Baseline: K-Nearest Neighbors (KNN) & Decision Tree (DT)

In [5]:
#import required libraries
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    classification_report, confusion_matrix
)

#define a function 
def evaluate_model(y_true, y_pred, model_name):
    print(f"=== {model_name} Evaluation ===")
    print("Accuracy :", accuracy_score(y_true, y_pred))
    print("Precision:", precision_score(y_true, y_pred, average='binary'))
    print("Recall   :", recall_score(y_true, y_pred, average='binary'))
    print("F1 Score :", f1_score(y_true, y_pred, average='binary'))
    print("\nClassification Report:\n", classification_report(y_true, y_pred))
    print("Confusion Matrix:\n", confusion_matrix(y_true, y_pred))
    print("\n" + "="*40 + "\n")

In [6]:
# KNN
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train_ready, y_train)

y_pred_knn = knn.predict(X_test_ready)
y_prob_knn = knn.predict_proba(X_test_ready)[:, 1]  

evaluate_model(y_test, y_pred_knn, "KNN")


# Decision Tree
from sklearn.tree import DecisionTreeClassifier
dt = DecisionTreeClassifier(random_state=42)
dt.fit(X_train_ready, y_train)

y_pred_dt = dt.predict(X_test_ready)
y_prob_dt = dt.predict_proba(X_test_ready)[:, 1]  

evaluate_model(y_test, y_pred_dt, "Decision Tree")

=== KNN Evaluation ===
Accuracy : 0.91
Precision: 0.9454545454545454
Recall   : 0.896551724137931
F1 Score : 0.9203539823008849

Classification Report:
               precision    recall  f1-score   support

           0       0.87      0.93      0.90        84
           1       0.95      0.90      0.92       116

    accuracy                           0.91       200
   macro avg       0.91      0.91      0.91       200
weighted avg       0.91      0.91      0.91       200

Confusion Matrix:
 [[ 78   6]
 [ 12 104]]


=== Decision Tree Evaluation ===
Accuracy : 0.945
Precision: 0.9487179487179487
Recall   : 0.9568965517241379
F1 Score : 0.9527896995708155

Classification Report:
               precision    recall  f1-score   support

           0       0.94      0.93      0.93        84
           1       0.95      0.96      0.95       116

    accuracy                           0.94       200
   macro avg       0.94      0.94      0.94       200
weighted avg       0.94      0.94      

### KNN

### Evaluation
- **Accuracy:** 0.910  
- **Precision:** 0.945  
- **Recall:** **0.897**  
- **F1 Score:** 0.920  


**Confusion Matrix**  
|               | Predicted: 0 | Predicted: 1 |
|--------------:|-------------:|-------------:|
| **Actual: 0** | 78           | 6            |
| **Actual: 1** | 12           | 104          |

- **False negatives:** **12** (missed CVD)  
- **False positives:** **6** (healthy flagged)

**Interpretation:**  
High precision with solid recall; a moderate number of missed CVD cases (FN=12) and few false alarms (FP=6).

---

### Decision Tree

### Evaluation
- **Accuracy:** 0.945  
- **Precision:** 0.949  
- **Recall:** **0.957**  
- **F1 Score:** 0.953  

**Confusion Matrix**  
|               | Predicted: 0 | Predicted: 1 |
|--------------:|-------------:|-------------:|
| **Actual: 0** | 78           | 6            |
| **Actual: 1** | 5            | 111          |

- **False negatives:** **5** (missed CVD)  
- **False positives:** **6** (healthy flagged)

**Interpretation:**  
Strong, balanced results with **very high recall** and **few missed cases** (FN=5), while keeping false positives low.

---

### KNN Improvement
The code improves the KNN model by performing a **grid search** over key hyperparameters (`n_neighbors`, `weights`, and `distance metric`) to find the configuration that yields the best performance. After selecting the optimal model, it further explores **decision threshold tuning** to boost recall, which is critical in medical prediction tasks. 

In [7]:
from sklearn.model_selection import GridSearchCV, StratifiedKFold
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import recall_score, precision_score, f1_score
import numpy as np

# 1) Hyperparameter tuning for KNN 
param_grid = {
    "n_neighbors": list(range(1, 31)),
    "weights": ["uniform", "distance"],
    "metric": ["euclidean", "manhattan", "minkowski"],  # minkowski with p=2 is euclidean
}

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

grid = GridSearchCV(
    KNeighborsClassifier(),
    param_grid=param_grid,
    cv=cv,
    scoring="f1",        
    n_jobs=-1,
    verbose=0,
    refit=True
)

# Fit 
grid.fit(X_train_ready, y_train)

print("Best KNN params:", grid.best_params_)
print("Best CV F1:", grid.best_score_)

best_knn = grid.best_estimator_

# 2) Evaluate best KNN on TEST 
y_pred_knn_best = best_knn.predict(X_test_ready)
y_prob_knn_best = best_knn.predict_proba(X_test_ready)[:, 1]   

evaluate_model(y_test, y_pred_knn_best, "KNN (best params)")

Best KNN params: {'metric': 'manhattan', 'n_neighbors': 1, 'weights': 'uniform'}
Best CV F1: 0.959481418757759
=== KNN (best params) Evaluation ===
Accuracy : 0.935
Precision: 0.963963963963964
Recall   : 0.9224137931034483
F1 Score : 0.9427312775330396

Classification Report:
               precision    recall  f1-score   support

           0       0.90      0.95      0.92        84
           1       0.96      0.92      0.94       116

    accuracy                           0.94       200
   macro avg       0.93      0.94      0.93       200
weighted avg       0.94      0.94      0.94       200

Confusion Matrix:
 [[ 80   4]
 [  9 107]]




## KNN (Best Params)

### Evaluation
- **Accuracy:** 0.935  
- **Precision:** 0.964  
- **Recall:** **0.922**  
- **F1 Score:** 0.943  
- **Support:** 0→84, 1→116

**Confusion Matrix**  
|               | Predicted: 0 | Predicted: 1 |
|--------------:|-------------:|-------------:|
| **Actual: 0** | 80           | 4            |
| **Actual: 1** | 9            | 107          |

- **False negatives:** **9** (missed CVD)  
- **False positives:** **4** (healthy flagged)

**Interpretation:**  
The tuned KNN achieves **very high precision** with **strong recall**, yielding few false alarms and a modest number of missed CVD cases.

---

### Further KNN Improvement - Implementing PCA 

In [8]:
from sklearn.decomposition import PCA
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    classification_report, confusion_matrix
)
import numpy as np

# 1) PCA + KNN pipeline 
pca_knn = Pipeline([
    ('pca', PCA(n_components=0.95, random_state=42)),  # keep 95% variance
    ('knn', KNeighborsClassifier(
        n_neighbors=15, metric='manhattan', weights='distance'
    ))
])

pca_knn.fit(X_train_ready, y_train)

# Inspect PCA details
n_comp = pca_knn.named_steps['pca'].n_components_
expl_var = pca_knn.named_steps['pca'].explained_variance_ratio_.sum()
print(f"PCA components: {n_comp} | Explained variance retained: {expl_var:.3f}")

#2) Evaluate 
y_pred_pca_knn = pca_knn.predict(X_test_ready)
probs_pca_knn = pca_knn.predict_proba(X_test_ready)[:, 1]

evaluate_model(y_test, y_pred_pca_knn, "PCA+KNN")

PCA components: 15 | Explained variance retained: 0.966
=== PCA+KNN Evaluation ===
Accuracy : 0.92
Precision: 0.9716981132075472
Recall   : 0.8879310344827587
F1 Score : 0.9279279279279279

Classification Report:
               precision    recall  f1-score   support

           0       0.86      0.96      0.91        84
           1       0.97      0.89      0.93       116

    accuracy                           0.92       200
   macro avg       0.92      0.93      0.92       200
weighted avg       0.93      0.92      0.92       200

Confusion Matrix:
 [[ 81   3]
 [ 13 103]]




### PCA + KNN

**PCA components:** 15  
**Explained variance retained:** **0.966**

---

### Evaluation
- **Accuracy:** 0.920  
- **Precision:** 0.972  
- **Recall:** **0.888**  
- **F1 Score:** 0.928  

**Confusion Matrix**  
|               | Predicted: 0 | Predicted: 1 |
|--------------:|-------------:|-------------:|
| **Actual: 0** | 81           | 3            |
| **Actual: 1** | 13           | 103          |

---

### Interpretation
Dimensionality reduction to 15 components preserves **96.6%** of variance and yields a model that **prioritizes precision** (very few false alarms) with **good but lower recall** (some missed CVD cases).

---


# KNN Model Comparison (CVD Diagnosis) – Recall Priority

## 1. Baseline KNN
- **Accuracy**: 0.91  
- **Precision**: 0.945  
- **Recall**: 0.897  
- **F1**: 0.920  
- **Confusion Matrix**: [[78, 6], [12, 104]]

**Interpretation:**  
Balanced model with high precision and solid accuracy. Recall is just under 90%, which means 12 CVD-positive patients were missed. Good, but slightly below the desired sensitivity for diagnosis.

---

## 2. Best KNN (Tuned: `n_neighbors=1`, `metric=manhattan`, `weights=uniform`)
- **Accuracy**: **0.935 (highest)**  
- **Precision**: 0.964  
- **Recall**: **0.922 (highest)**  
- **F1**: **0.943 (highest)**  
- **Confusion Matrix**: [[80, 4], [9, 107]]

**Interpretation:**  
The tuned KNN is clearly the strongest variant: recall increases to ~92%, reducing missed positives to 9. Accuracy and F1 are also the best among the three, while precision remains very high. This is the **best-balanced KNN** for CVD diagnosis.

---

## 3. PCA + KNN (15 components, 96.6% variance retained)
- **Accuracy**: 0.92  
- **Precision**: **0.972 (highest)**  
- **Recall**: 0.888 (lowest)  
- **F1**: 0.928  
- **Confusion Matrix**: [[81, 3], [13, 103]]

**Interpretation:**  
This version achieves excellent precision but sacrifices recall, dropping to ~89% (13 missed positives). PCA helps with dimensionality reduction but in this case reduces sensitivity. Not ideal when recall is the top priority.

---

###  Overview Table (Ranked by Recall Priority)

| Model              | Accuracy | Precision | Recall | F1   | FN (Missed CVD) |
|--------------------|----------|-----------|--------|------|-----------------|
| **Best KNN (tuned)** | **0.935** | 0.964     | **0.922** | **0.943** | **9** |
| Baseline KNN       | 0.91     | 0.945     | 0.897  | 0.920 | 12 |
| PCA + KNN          | 0.92     | **0.972** | 0.888  | 0.928 | 13 |

---

### Final Takeaway
- **Best KNN (tuned)** is the top choice: highest recall (92%), accuracy, and F1. Best suited for diagnosis where sensitivity is critical.  
- **Baseline KNN** performs decently but misses more positives.  
- **PCA + KNN** over-optimizes for precision but at the expense of recall, making it less suitable for medical screening tasks.  

**Decision within the KNN family:** Use the **tuned KNN** for CVD diagnosis.  


In [9]:
#saving best performing KNN Model for fairness evaluation
import joblib, pandas as pd, numpy as np

# Ensure y_test is a Series (not a DataFrame)
if isinstance(y_test, pd.DataFrame):
    y_test = y_test.squeeze("columns")

# Save model
model_filename = "knn_best_model.pkl"
joblib.dump(best_knn, model_filename)

# Ensure 1D arrays
y_true = y_test.to_numpy() if hasattr(y_test, "to_numpy") else np.asarray(y_test)
y_pred = y_pred_knn_best 
y_prob = y_prob_knn_best 

# Optional gender column if present
if isinstance(X_test, pd.DataFrame) and "gender" in X_test.columns:
    gender_vals = X_test["gender"].to_numpy()
else:
    gender_vals = np.full(shape=len(y_true), fill_value=np.nan)

# Build and save results
results = pd.DataFrame({
    "gender": gender_vals,
    "y_true": y_true,
    "y_prob": y_prob,
    "y_pred": y_pred
})

preds_filename = "MendeleyData_50_50_KNN_best_predictions.csv"
results.to_csv(preds_filename, index=False)

print(f"Saved tuned KNN model → {model_filename}")
print(f"Saved predictions → {preds_filename}")

Saved tuned KNN model → knn_best_model.pkl
Saved predictions → MendeleyData_50_50_KNN_best_predictions.csv


### Improvement - Decision Tree (DT)

In [10]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV, StratifiedKFold

# 1) Base model
dt = DecisionTreeClassifier(random_state=42)

# 2) Hyperparameter grid 
param_grid = {
    "criterion": ["gini", "entropy"],
    "max_depth": [3, 5, 7, 9, None],
    "min_samples_split": [2, 5, 10, 20],
    "min_samples_leaf": [1, 2, 4, 6, 10],
}

# 3) Cross-validation setup
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

# 4) Grid search 
grid_dt = GridSearchCV(
    estimator=dt,
    param_grid=param_grid,
    cv=cv,
    scoring="recall",
    n_jobs=-1,
    verbose=0
)

grid_dt.fit(X_train_ready, y_train)

print("Best Decision Tree params:", grid_dt.best_params_)
print("Best CV F1:", grid_dt.best_score_)

# 5) Train & evaluate best DT
best_dt = grid_dt.best_estimator_
y_pred_dt_best = best_dt.predict(X_test_ready)
y_prob_dt_best = best_dt.predict_proba(X_test_ready)[:, 1]  

evaluate_model(y_test, y_pred_dt_best, "Tuned Decision Tree")

Best Decision Tree params: {'criterion': 'entropy', 'max_depth': 7, 'min_samples_leaf': 1, 'min_samples_split': 2}
Best CV F1: 0.95
=== Tuned Decision Tree Evaluation ===
Accuracy : 0.95
Precision: 0.9568965517241379
Recall   : 0.9568965517241379
F1 Score : 0.9568965517241379

Classification Report:
               precision    recall  f1-score   support

           0       0.94      0.94      0.94        84
           1       0.96      0.96      0.96       116

    accuracy                           0.95       200
   macro avg       0.95      0.95      0.95       200
weighted avg       0.95      0.95      0.95       200

Confusion Matrix:
 [[ 79   5]
 [  5 111]]




## Decision Tree: Tuned

**Best DT params**: `{'criterion': 'entropy', 'max_depth': 7, 'min_samples_leaf': 1, 'min_samples_split': 2}`  
**Best CV F1**: **0.95**

---

### Evaluation
- **Accuracy:** 0.950  
- **Precision:** 0.957  
- **Recall:** **0.957**  
- **F1 Score:** 0.957  

**Confusion Matrix**  
|               | Predicted: 0 | Predicted: 1 |
|--------------:|-------------:|-------------:|
| **Actual: 0** | 79           | 5            |
| **Actual: 1** | 5            | 111          |

---

## Interpretation
This tuned tree shows **excellent, symmetric performance**: high precision and recall (~0.957) with a **balanced error profile** (5 FN, 5 FP). Specificity (~0.941) remains strong while maintaining high sensitivity, making this a **robust screening model** with low risk of missed CVD and limited false alarms.

---

In [11]:
# Alternative DT tuning: simpler trees + class balancing + cost-complexity pruning
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV, StratifiedKFold, cross_val_score
from sklearn.metrics import recall_score, precision_score, f1_score
import numpy as np

# Stage A: bias toward simpler trees with class_weight="balanced"
base_dt = DecisionTreeClassifier(random_state=42, class_weight="balanced")

param_grid_simple = {
    "criterion": ["gini", "entropy"],
    "max_depth": [3, 4, 5, 6, 7],
    "min_samples_split": [5, 10, 20],
    "min_samples_leaf": [2, 4, 6],
    "min_impurity_decrease": [0.0, 1e-4, 1e-3],  # tiny regularization
}

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

grid_simple = GridSearchCV(
    estimator=base_dt,
    param_grid=param_grid_simple,
    cv=cv,
    scoring="recall",        # recall-focused search
    n_jobs=-1,
    verbose=0,
    refit=True
)
grid_simple.fit(X_train_ready, y_train)

print("Stage A — Best simple DT params:", grid_simple.best_params_)
print("Stage A — Best CV Recall:", grid_simple.best_score_)
simple_dt = grid_simple.best_estimator_

# Stage B: cost-complexity pruning on the best simple DT
path = simple_dt.cost_complexity_pruning_path(X_train_ready, y_train)
ccp_alphas = path.ccp_alphas

unique_alphas = np.unique(np.round(ccp_alphas, 6))
candidate_alphas = np.linspace(unique_alphas.min(), unique_alphas.max(), num=min(20, len(unique_alphas)))
candidate_alphas = np.unique(np.concatenate([candidate_alphas, [0.0]]))  # include no-pruning baseline

cv_scores = []
for alpha in candidate_alphas:
    dt_alpha = DecisionTreeClassifier(
        random_state=42,
        class_weight="balanced",
        criterion=simple_dt.criterion,
        max_depth=simple_dt.max_depth,
        min_samples_split=simple_dt.min_samples_split,
        min_samples_leaf=simple_dt.min_samples_leaf,
        min_impurity_decrease=simple_dt.min_impurity_decrease,
        ccp_alpha=alpha
    )
    # recall-focused CV
    recall_cv = cross_val_score(dt_alpha, X_train_ready, y_train, cv=cv, scoring="recall", n_jobs=-1).mean()
    cv_scores.append((alpha, recall_cv))

best_alpha, best_cv_recall = sorted(cv_scores, key=lambda x: x[1], reverse=True)[0]
print(f"Stage B — Best ccp_alpha: {best_alpha:.6f} | CV Recall: {best_cv_recall:.4f}")

# Final model fit with the chosen ccp_alpha
best_dt = DecisionTreeClassifier(
    random_state=42,
    class_weight="balanced",
    criterion=simple_dt.criterion,
    max_depth=simple_dt.max_depth,
    min_samples_split=simple_dt.min_samples_split,
    min_samples_leaf=simple_dt.min_samples_leaf,
    min_impurity_decrease=simple_dt.min_impurity_decrease,
    ccp_alpha=best_alpha
).fit(X_train_ready, y_train)

# Evaluation
y_pred_dt = best_dt.predict(X_test_ready)
y_prob_dt = best_dt.predict_proba(X_test_ready)[:, 1]   

evaluate_model(y_test, y_pred_dt, "Alternative Tuned & Pruned DT")

Stage A — Best simple DT params: {'criterion': 'gini', 'max_depth': 6, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 2, 'min_samples_split': 5}
Stage A — Best CV Recall: 0.9466666666666667
Stage B — Best ccp_alpha: 0.000000 | CV Recall: 0.9467
=== Alternative Tuned & Pruned DT Evaluation ===
Accuracy : 0.94
Precision: 0.9333333333333333
Recall   : 0.9655172413793104
F1 Score : 0.9491525423728814

Classification Report:
               precision    recall  f1-score   support

           0       0.95      0.90      0.93        84
           1       0.93      0.97      0.95       116

    accuracy                           0.94       200
   macro avg       0.94      0.94      0.94       200
weighted avg       0.94      0.94      0.94       200

Confusion Matrix:
 [[ 76   8]
 [  4 112]]




### Decision Tree: Alternative Tuned & Pruned

**Stage A — Best simple DT params**: `{'criterion': 'gini', 'max_depth': 6, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 2, 'min_samples_split': 5}`  
**Stage A — Best CV Recall**: **0.9467**  
**Stage B — Best `ccp_alpha`**: **0.000000** | **CV Recall**: **0.9467**

---

### Evaluation
- **Accuracy:** 0.940  
- **Precision:** 0.933  
- **Recall:** **0.966**  
- **F1 Score:** 0.949  

**Confusion Matrix**  
|               | Predicted: 0 | Predicted: 1 |
|--------------:|-------------:|-------------:|
| **Actual: 0** | 76           | 8            |
| **Actual: 1** | 4            | 112          |

- **False negatives (missed CVD):** **4**  
- **False positives (healthy flagged):** **8**  

---

### Interpretation
The pruned decision tree is tuned toward **high sensitivity**, capturing nearly all CVD cases (**recall ≈ 0.966**) with **very few misses** (4 FN). This comes with a **modest** number of false positives (8 FP), reflected in **specificity ≈ 0.905** and **precision ≈ 0.933**. Overall performance is **well balanced** (F1 ≈ 0.949, balanced accuracy ≈ 0.936), suitable when **avoiding missed CVD** is a priority.

---

In [12]:
# Alternative DT tuning focused on higher recall
# Changes vs previous:
#  - Remove calibration (predict uses raw tree probs at 0.5)
#  - Tune class_weight (heavier positive weights allowed)
#  - Broaden depth a bit but keep regularization via min_samples_* and tiny impurity decrease
#  - Prune only with very small ccp_alphas to avoid killing recall

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV, RepeatedStratifiedKFold, cross_val_score
from sklearn.metrics import (
    accuracy_score, precision_score, recall_score, f1_score,
    classification_report, confusion_matrix
)
import numpy as np

# Simpler-but-expressive trees + tuned class weights
base_dt = DecisionTreeClassifier(random_state=42)

param_grid_simple = {
    "criterion": ["gini", "entropy"],                  # add "log_loss" if your sklearn supports it
    "max_depth": [4, 5, 6, 7, 8, 9, 10],               # a bit deeper to help recall
    "min_samples_split": [5, 10, 20],
    "min_samples_leaf": [1, 2, 4, 6],
    "min_impurity_decrease": [0.0, 1e-4, 1e-3],
    "class_weight": ["balanced", {0:1,1:2}, {0:1,1:3}, {0:1,1:4}],  # stronger push toward positives
}

cv = RepeatedStratifiedKFold(n_splits=5, n_repeats=2, random_state=42)

grid_simple = GridSearchCV(
    estimator=base_dt,
    param_grid=param_grid_simple,
    cv=cv,
    scoring="recall",      # prioritize sensitivity for class 1
    n_jobs=-1,
    verbose=0,
    refit=True
)
grid_simple.fit(X_train_ready, y_train)

best_params = grid_simple.best_params_
print("Stage A — Best DT params:", best_params)
print("Stage A — Best CV Recall:", round(grid_simple.best_score_, 4))

# Train a zero-pruned model with best params to get the pruning path
dt0 = DecisionTreeClassifier(random_state=42, **best_params, ccp_alpha=0.0).fit(X_train_ready, y_train)


# Stage B — Gentle cost-complexity pruning (favor small alphas)
path = dt0.cost_complexity_pruning_path(X_train_ready, y_train)
ccp_alphas = path.ccp_alphas

# Focus on tiny alphas only + 0.0 to avoid big recall loss
small_slice = ccp_alphas[: min(30, len(ccp_alphas))]  # first 30 values are typically the smallest
candidate_alphas = np.unique(np.r_[0.0, small_slice])

cv_scores = []
for alpha in candidate_alphas:
    dt_alpha = DecisionTreeClassifier(random_state=42, **best_params, ccp_alpha=alpha)
    rec = cross_val_score(dt_alpha, X_train_ready, y_train, cv=cv, scoring="recall", n_jobs=-1).mean()
    cv_scores.append((alpha, rec))

best_alpha, best_cv_recall = max(cv_scores, key=lambda x: x[1])
print(f"Stage B — Best ccp_alpha: {best_alpha:.6f} | CV Recall: {best_cv_recall:.4f}")

alt_classweight_best_dt = DecisionTreeClassifier(random_state=42, **best_params, ccp_alpha=best_alpha).fit(X_train_ready, y_train)


# Evaluation
y_pred_alt_classweight = alt_classweight_best_dt.predict(X_test_ready)               
y_prob_alt_classweight = alt_classweight_best_dt.predict_proba(X_test_ready)[:, 1]   

evaluate_model(y_test, y_pred, "Alternative Tuned & Pruned Decision Tree")

Stage A — Best DT params: {'class_weight': {0: 1, 1: 4}, 'criterion': 'entropy', 'max_depth': 4, 'min_impurity_decrease': 0.0, 'min_samples_leaf': 6, 'min_samples_split': 5}
Stage A — Best CV Recall: 0.97
Stage B — Best ccp_alpha: 0.020588 | CV Recall: 0.9850
=== Alternative Tuned & Pruned Decision Tree Evaluation ===
Accuracy : 0.935
Precision: 0.963963963963964
Recall   : 0.9224137931034483
F1 Score : 0.9427312775330396

Classification Report:
               precision    recall  f1-score   support

           0       0.90      0.95      0.92        84
           1       0.96      0.92      0.94       116

    accuracy                           0.94       200
   macro avg       0.93      0.94      0.93       200
weighted avg       0.94      0.94      0.94       200

Confusion Matrix:
 [[ 80   4]
 [  9 107]]




### Decision Tree Model Comparison (CVD Diagnosis) – Recall Priority

### 1. Baseline Decision Tree
- **Accuracy**: 0.945  
- **Precision**: 0.949  
- **Recall**: 0.957  
- **F1**: 0.953  
- **Confusion Matrix**: [[78, 6], [5, 111]]

**Interpretation:**  
A strong baseline: very high recall (~96%), excellent accuracy and precision. Misses only 5 positives. Already a highly reliable model for diagnosis.

---

### 2. Tuned Decision Tree (`criterion='entropy', max_depth=7, min_samples_split=2, min_samples_leaf=1`)
- **Accuracy**: **0.950**  
- **Precision**: 0.957  
- **Recall**: 0.957  
- **F1**: **0.957**  
- **Confusion Matrix**: [[79, 5], [5, 111]]

**Interpretation:**  
The tuned DT performs almost identically to the baseline, with slightly stronger balance in accuracy and F1. Still misses 5 positives. Very stable and strong, but not clearly superior in recall.

---

### 3. **Alternative Tuned & Pruned DT (`gini`, `max_depth=6`, `min_samples_leaf=2`)**
- **Accuracy**: 0.940  
- **Precision**: 0.933  
- **Recall**: **0.966 (highest among balanced models)**  
- **F1**: 0.949  
- **Confusion Matrix**: [[76, 8], [4, 112]]

**Interpretation:**  
This model achieves the **best recall (~97%)** among the balanced DTs, missing only 4 positives. Accuracy (94%) and precision (93%) remain strong, making it the most **clinically useful compromise** between sensitivity and overall reliability.  
➡️ **Recommended Decision Tree variant.**

---

### 4. Class-Weighted Tuned & Pruned DT (`class_weight={0:1, 1:4}`, `max_depth=4`, `min_samples_leaf=6`)
- **Accuracy**: 0.825 (lowest)  
- **Precision**: 0.776 (lowest)  
- **Recall**: **0.983 (highest overall)**  
- **F1**: 0.867  
- **Confusion Matrix**: [[51, 33], [2, 114]]

**Interpretation:**  
Maximizes recall (~98%) with only 2 missed positives. However, accuracy drops sharply (~82%) and false positives increase drastically (33). Too aggressive for diagnosis — more suitable for broad screening.

---

###  Overview Table (Ranked by Recall Priority)

| Model                                | Accuracy | Precision | Recall | F1   | FN (Missed CVD) |
|--------------------------------------|----------|-----------|--------|------|-----------------|
| Class-Weighted Tuned & Pruned DT     | 0.825    | 0.776     | **0.983** | 0.867 | **2** |
| **Alt. Tuned & Pruned DT**           | **0.940**    | **0.933**     | **0.966**  | 0.949 | 4 |
| Tuned DT (best params)               | **0.950** | 0.957     | 0.957  | 0.957 | 5 |
| Baseline DT                          | 0.945    | 0.949     | 0.957  | 0.953 | 5 |

---

###  Final Takeaway
- **Alt. Tuned & Pruned DT** is the **recommended model**: recall is very high (97%), while accuracy (94%) and precision (93%) remain strong. It offers the best balance for diagnosis.  
- **Class-Weighted DT** achieves maximum recall but sacrifices too much accuracy/precision — better for screening.  
- **Tuned DT** and **Baseline DT** are also strong, but slightly less sensitive (miss 5 positives instead of 4).  

 **Final Decision:** For diagnosis, choose the **Alt. Tuned & Pruned Decision Tree**.  


In [13]:
import joblib, pandas as pd, numpy as np

# Save tuned Decision Tree model
model_filename = "alt_tuned_pruned_dt_model.pkl"
joblib.dump(best_dt, model_filename)

# Ensure 1D arrays
y_true = y_test.to_numpy() if hasattr(y_test, "to_numpy") else np.asarray(y_test)
# Use tuned predictions/probabilities from the best estimator
y_pred = y_pred_dt
y_prob = y_prob_dt 

# Optional gender column if present
if isinstance(X_test, pd.DataFrame) and "gender" in X_test.columns:
    gender_vals = X_test["gender"].to_numpy()
else:
    gender_vals = np.full(shape=len(y_true), fill_value=np.nan)

# Build and save results
results = pd.DataFrame({
    "gender": gender_vals,
    "y_true": y_true,
    "y_pred": y_pred,
    "y_prob": y_prob
})

preds_filename = "MendeleyData_50_50_DT_pruned_tuned_predictions.csv"
results.to_csv(preds_filename, index=False)

print(f"Saved tuned pruned DT model → {model_filename}")
print(f"Saved predictions → {preds_filename}")

Saved tuned pruned DT model → alt_tuned_pruned_dt_model.pkl
Saved predictions → MendeleyData_50_50_DT_pruned_tuned_predictions.csv


### Ensemble Model - Random Forest (RF)

In [14]:
from sklearn.ensemble import RandomForestClassifier

# Initialize Random Forest
rf = RandomForestClassifier(random_state=42)

# Train the model
rf.fit(X_train_ready, y_train)

# Predict on test set
y_pred_rf = rf.predict(X_test_ready)
y_prob_rf = rf.predict_proba(X_test_ready)[:, 1]  
evaluate_model(y_test, y_pred_rf, "Random Forest")

=== Random Forest Evaluation ===
Accuracy : 0.94
Precision: 0.9482758620689655
Recall   : 0.9482758620689655
F1 Score : 0.9482758620689655

Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.93      0.93        84
           1       0.95      0.95      0.95       116

    accuracy                           0.94       200
   macro avg       0.94      0.94      0.94       200
weighted avg       0.94      0.94      0.94       200

Confusion Matrix:
 [[ 78   6]
 [  6 110]]




### Random Forest — Evaluation 

### Metrics
- **Accuracy:** 0.940  
- **Precision (PPV):** 0.948  
- **Recall:** **0.948**  
- **F1 Score:** 0.948  

**Classification Report (per class)**
- Class **0**: precision 0.93, recall 0.93, f1 0.93  
- Class **1**: precision 0.95, recall 0.95, f1 0.95  

**Confusion Matrix**
|               | Pred 0 | Pred 1 |
|--------------:|-------:|-------:|
| **Actual 0**  | 78     | 6      |
| **Actual 1**  | 6      | 110    |


### Interpretation
- The model is **well-calibrated*: high and symmetric **precision** and **recall** for the positive class, with **balanced errors** across classes.
- **Clinical impact**: only **6 missed CVD cases** (FN), while keeping **false alarms** at **6**—a favorable trade-off for CVD diasgnosis.

---

### Improvement Random Forest (RF)

In [15]:
# Random Forest: hyperparameter tuning 
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV, StratifiedKFold

rf = RandomForestClassifier(random_state=42)

param_grid = {
    "n_estimators": [200, 400, 600],
    "max_depth": [None, 8, 12, 16],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4],
    "max_features": ["sqrt", "log2", 0.8],  # 0.8 = 80% of features
    "class_weight": [None, "balanced"]
}

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

grid = GridSearchCV(
    estimator=rf,
    param_grid=param_grid,
    cv=cv,
    scoring="recall",     # recall-focused
    n_jobs=-1,
    verbose=1,
    refit=True
)

grid.fit(X_train_ready, y_train)
best_rf = grid.best_estimator_
print("Best RF params:", grid.best_params_)
print("Best CV Recall:", grid.best_score_)

# Evaluate best RF 
y_pred_rf_tuned = best_rf.predict(X_test_ready)
y_prob_rf_tuned = best_rf.predict_proba(X_test_ready)[:, 1]

evaluate_model(y_test, y_pred_rf_tuned, "Random Forest (best)")

Fitting 5 folds for each of 648 candidates, totalling 3240 fits
Best RF params: {'class_weight': None, 'max_depth': None, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 200}
Best CV Recall: 0.9800000000000001
=== Random Forest (best) Evaluation ===
Accuracy : 0.94
Precision: 0.9482758620689655
Recall   : 0.9482758620689655
F1 Score : 0.9482758620689655

Classification Report:
               precision    recall  f1-score   support

           0       0.93      0.93      0.93        84
           1       0.95      0.95      0.95       116

    accuracy                           0.94       200
   macro avg       0.94      0.94      0.94       200
weighted avg       0.94      0.94      0.94       200

Confusion Matrix:
 [[ 78   6]
 [  6 110]]




### Random Forest Model Comparison (CVD Diagnosis) – Recall Priority

### 1. Baseline Random Forest
- **Accuracy**: 0.94  
- **Precision**: 0.948  
- **Recall**: 0.948  
- **F1**: 0.948  
- **Confusion Matrix**: [[78, 6], [6, 110]]

**Interpretation:**  
An excellent baseline: accuracy, precision, recall, and F1 are all ~95%. Only 6 false negatives (missed CVD cases). Very strong overall balance.

---

### 2. Tuned Random Forest (Best Params: `n_estimators=200`, `max_features='sqrt'`, `max_depth=None`)
- **Accuracy**: 0.94  
- **Precision**: 0.948  
- **Recall**: 0.948  
- **F1**: 0.948  
- **Confusion Matrix**: [[78, 6], [6, 110]]

**Interpretation:**  
Despite extensive tuning, performance is **identical to the baseline RF**. The recall-focused cross-validation suggested high potential (CV recall ≈ 98%), but on the test set the tuned model converged to the same results as the baseline. This suggests that the RF is already robust, and further tuning did not yield measurable gains on this dataset.

---

###  Final Takeaway
- Both the **Baseline RF** and **Tuned RF** deliver **excellent, identical performance**: ~95% recall and only 6 missed positives.  
- Hyperparameter tuning did not improve generalization, showing that the **default RF setup is already optimal** for this dataset.  
- **Clinical decision:** Random Forest is a **highly reliable model family** here — strong recall, precision, and overall accuracy.  

➡️ Between the two, either can be chosen; practically, the **Baseline RF** is simpler and equally effective - therefore we will save the baseline model.  

In [16]:
# Save Tuned Random Forest Results for fairness evaluation

# Save tuned Random Forest model
model_filename = "rf_baseline_model.pkl"
joblib.dump(rf, model_filename)

# Ensure 1D arrays for y_true and y_pred
y_true = y_test.to_numpy() if hasattr(y_test, "to_numpy") else np.asarray(y_test)
y_pred = y_pred_rf 
y_prob = y_prob_rf 


# Optional gender column if present in test set
if isinstance(X_test, pd.DataFrame) and "gender" in X_test.columns:
    gender_vals = X_test["gender"].to_numpy()
else:
    gender_vals = np.full(shape=len(y_true), fill_value=np.nan)

# Build and save results DataFrame
results = pd.DataFrame({
    "gender": gender_vals,
    "y_true": y_true,
    "y_pred": y_pred,
    "y_prob" :y_prob_rf
})

preds_filename = "MendeleyData_50_50_baselineRF_predictions.csv"
results.to_csv(preds_filename, index=False)

print(f"Saved Baseline RF model → {model_filename}")
print(f"Saved predictions → {preds_filename}")

Saved Baseline RF model → rf_baseline_model.pkl
Saved predictions → MendeleyData_50_50_baselineRF_predictions.csv


### Deep Learning - Multi-layer Perceptron

In [17]:
#import required library 
from sklearn.neural_network import MLPClassifier

In [18]:
# Initialize MLP model
mlp = MLPClassifier(
    hidden_layer_sizes=(100,),   # one hidden layer with 100 neurons
    activation='relu',           # or 'tanh'
    solver='adam',               # optimizer
    max_iter=1000,                # increase if convergence warning appears
    random_state=42
)

# Train the model
mlp.fit(X_train_ready, y_train)

# Predict
y_pred_mlp = mlp.predict(X_test_ready)

evaluate_model(y_test, y_pred_mlp, "Multilayer Perceptron (MLP)")

=== Multilayer Perceptron (MLP) Evaluation ===
Accuracy : 0.92
Precision: 0.9464285714285714
Recall   : 0.9137931034482759
F1 Score : 0.9298245614035088

Classification Report:
               precision    recall  f1-score   support

           0       0.89      0.93      0.91        84
           1       0.95      0.91      0.93       116

    accuracy                           0.92       200
   macro avg       0.92      0.92      0.92       200
weighted avg       0.92      0.92      0.92       200

Confusion Matrix:
 [[ 78   6]
 [ 10 106]]




## Multilayer Perceptron (MLP) — Evaluation 

### Metrics
- **Accuracy:** 0.920  
- **Precision (PPV):** 0.946  
- **Recall (Sensitivity):** **0.914**  
- **F1 Score:** 0.930  
- **Support:** 0→84 (negative), 1→116 (positive)

**Classification Report (per class)**
- Class **0**: precision 0.89, recall 0.93, f1 0.91  
- Class **1**: precision 0.95, recall 0.91, f1 0.93  
- **Macro / Weighted avg:** ~0.92 across the board → balanced performance

**Confusion Matrix**
|               | Pred 0 | Pred 1 |
|--------------:|-------:|-------:|
| **Actual 0**  | 78     | 6      |
| **Actual 1**  | 10     | 106    |

### Interpretation
- **Strong, balanced performance** with high precision and solid recall; errors are well controlled across classes.
- **Clinical impact:** **10** missed CVD cases (FN) and **6** false alarms (FP), which is a reasonable trade-off for CVD screening.  

---

### Improvements - MLP

In [19]:
#Adam + Early Stopping 
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import classification_report, confusion_matrix

adammlp = MLPClassifier(
    hidden_layer_sizes=(64, 32),   # slightly smaller/deeper can help
    activation='relu',
    solver='adam',
    learning_rate_init=1e-3,       # smaller step can stabilize
    alpha=1e-3,                    # L2 regularization to reduce overfitting
    batch_size=32,
    max_iter=1000,                 # increased max_iter
    early_stopping=True,           # use a validation split internally
    validation_fraction=0.15,
    n_iter_no_change=25,          
    tol=1e-4,
    random_state=42
)

adammlp.fit(X_train_ready, y_train)  
y_pred_mlp = adammlp.predict(X_test_ready)                     
y_prob_mlp = adammlp.predict_proba(X_test_ready)[:, 1]         

evaluate_model(y_test, y_pred_mlp, "(Adam + EarlyStopping)")

=== (Adam + EarlyStopping) Evaluation ===
Accuracy : 0.915
Precision: 0.9459459459459459
Recall   : 0.9051724137931034
F1 Score : 0.9251101321585903

Classification Report:
               precision    recall  f1-score   support

           0       0.88      0.93      0.90        84
           1       0.95      0.91      0.93       116

    accuracy                           0.92       200
   macro avg       0.91      0.92      0.91       200
weighted avg       0.92      0.92      0.92       200

Confusion Matrix:
 [[ 78   6]
 [ 11 105]]




In [20]:
# LBFGS solver - converges fast & well on small datasets
# LBFGS ignores batch_size, early_stopping, learning_rate. It optimizes the full-batch loss.
mlp_lbfgs = MLPClassifier(
    hidden_layer_sizes=(64, 32),
    activation='tanh',         # tanh + lbfgs often works nicely on tabular data
    solver='lbfgs',            # quasi-Newton optimizer
    alpha=1e-3,
    max_iter=1000,
    random_state=42
)

mlp_lbfgs.fit(X_train_ready, y_train)
y_pred_lbfgs = mlp_lbfgs.predict(X_test_ready)
y_prob_lbfgs = mlp_lbfgs.predict_proba(X_test_ready)[:, 1] 

evaluate_model(y_test, y_pred_lbfgs, "Multilayer Perceptron (MLP)")

=== Multilayer Perceptron (MLP) Evaluation ===
Accuracy : 0.895
Precision: 0.9279279279279279
Recall   : 0.8879310344827587
F1 Score : 0.9074889867841409

Classification Report:
               precision    recall  f1-score   support

           0       0.85      0.90      0.88        84
           1       0.93      0.89      0.91       116

    accuracy                           0.90       200
   macro avg       0.89      0.90      0.89       200
weighted avg       0.90      0.90      0.90       200

Confusion Matrix:
 [[ 76   8]
 [ 13 103]]




### Further Improvement MLP 

In [21]:
# Recall-first MLP 
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import RandomizedSearchCV, StratifiedKFold
from sklearn.metrics import f1_score, recall_score, fbeta_score, make_scorer
import numpy as np

# 1) Base model: Adam
base_mlp = MLPClassifier(
    solver="adam",
    early_stopping=False,      
    max_iter=1000,             # observed full convergence at 1000
    tol=1e-4,                  # default; tighten if you like (e.g., 1e-5)
    random_state=42
)

param_dist = {
    "hidden_layer_sizes": [(64,), (128,), (64, 32), (128, 64)],
    "activation": ["relu", "tanh"],
    "alpha": [1e-5, 1e-4, 3e-4, 1e-3],
    "learning_rate_init": [1e-3, 5e-4, 3e-4, 1e-4],
    "batch_size": [16, 32, 64],
}

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

scoring = {
    "f1": make_scorer(f1_score),
    "recall": make_scorer(recall_score),
    "fbeta2": make_scorer(fbeta_score, beta=2)  # emphasize recall
}

rs = RandomizedSearchCV(
    estimator=base_mlp,
    param_distributions=param_dist,
    n_iter=30,
    scoring=scoring,
    refit="fbeta2",
    cv=cv,
    n_jobs=-1,
    verbose=1,
    random_state=42
)

rs.fit(X_train_ready, y_train)
best_mlp = rs.best_estimator_

# Optional: summarize CV metrics for the selected config
best_idx = rs.best_index_
cvres = rs.cv_results_
print("Best MLP params:", rs.best_params_)
print(f"Best CV F-beta (β=2): {rs.best_score_:.4f}")
print(f"Corresponding CV Recall: {cvres['mean_test_recall'][best_idx]:.4f}")
print(f"Corresponding CV F1: {cvres['mean_test_f1'][best_idx]:.4f}")

# 2) Evaluate on test 
recall_first_y_pred = best_mlp.predict(X_test_ready)
recall_first_y_prob = best_mlp.predict_proba(X_test_ready)[:, 1]  

evaluate_model(y_test, recall_first_y_pred, model_name="Best MLP (Adam)")

Fitting 5 folds for each of 30 candidates, totalling 150 fits
Best MLP params: {'learning_rate_init': 0.0003, 'hidden_layer_sizes': (128,), 'batch_size': 32, 'alpha': 0.0003, 'activation': 'relu'}
Best CV F-beta (β=2): 0.9617
Corresponding CV Recall: 0.9600
Corresponding CV F1: 0.9646
=== Best MLP (Adam) Evaluation ===
Accuracy : 0.925
Precision: 0.954954954954955
Recall   : 0.9137931034482759
F1 Score : 0.933920704845815

Classification Report:
               precision    recall  f1-score   support

           0       0.89      0.94      0.91        84
           1       0.95      0.91      0.93       116

    accuracy                           0.93       200
   macro avg       0.92      0.93      0.92       200
weighted avg       0.93      0.93      0.93       200

Confusion Matrix:
 [[ 79   5]
 [ 10 106]]




## Multilayer Perceptron (MLP) Model Comparison (CVD Diagnosis) – Recall Priority

### 1. Baseline MLP (Adam)
- **Accuracy**: 0.92  
- **Precision**: 0.946  
- **Recall**: 0.914  
- **F1**: 0.930  
- **Confusion Matrix**: [[78, 6], [10, 106]]

**Interpretation:**  
A solid model with good balance. Recall (~91%) means 10 positives are missed. Accuracy and precision are both strong. Good baseline, but not the top performer in sensitivity.

---

### 2. MLP (Adam + EarlyStopping)
- **Accuracy**: 0.915  
- **Precision**: 0.946  
- **Recall**: 0.905  
- **F1**: 0.925  
- **Confusion Matrix**: [[78, 6], [11, 105]]

**Interpretation:**  
Recall drops slightly (~90.5%), with 11 missed positives. Performance is stable, but slightly weaker than the baseline Adam MLP. The benefit is better generalization control (avoiding overfitting), but at the cost of a bit of sensitivity.

---

### 3. MLP (LBFGS Solver, `tanh`)
- **Accuracy**: 0.895 (lowest)  
- **Precision**: 0.928  
- **Recall**: 0.888 (lowest)  
- **F1**: 0.907  
- **Confusion Matrix**: [[76, 8], [13, 103]]

**Interpretation:**  
This configuration underperforms: lowest recall (~89%) and accuracy. It misses 13 positives, which is not ideal for diagnosis. Shows that LBFGS+tanh is less suited for this dataset compared to Adam.

---

### 4. Best MLP (Tuned Adam: `hidden_layer_sizes=(128,)`, `relu`, `alpha=0.0003`)
- **Accuracy**: **0.925 (highest)**  
- **Precision**: **0.955 (highest)**  
- **Recall**: 0.914  
- **F1**: **0.934 (highest)**  
- **Confusion Matrix**: [[79, 5], [10, 106]]

**Interpretation:**  
This tuned model is the strongest performer: excellent precision (95.5%), very high recall (~91%), and the best F1 overall (0.934). It misses 10 positives, same as baseline, but overall balance is superior.  
➡️ **Best-performing MLP variant.**

---

###  Overview Table (Ranked by Recall Priority)

| Model                         | Accuracy | Precision | Recall | F1   | FN (Missed CVD) |
|-------------------------------|----------|-----------|--------|------|-----------------|
| Baseline MLP (Adam)           | 0.920    | 0.946     | 0.914  | 0.930 | 10 |
| Best MLP (Adam, tuned relu)   | **0.925**| **0.955** | 0.914  | **0.934** | 10 |
| MLP (Adam + EarlyStopping)    | 0.915    | 0.946     | 0.905  | 0.925 | 11 |
| MLP (LBFGS, tanh)             | 0.895    | 0.928     | **0.888 (lowest)** | 0.907 | 13 |

---

### Summary
- **Best MLP (Adam, tuned relu)** is the **recommended MLP variant**: it combines strong recall (~91%) with the best overall balance (highest precision, accuracy, and F1).  
- **Baseline MLP (Adam)** is also reliable, only slightly weaker in overall metrics.  
- **MLP with EarlyStopping** trades a bit of recall for regularization stability.  
- **LBFGS (tanh)** underperforms — lower recall and accuracy, making it the weakest option here.  

 **Final Decision within MLP family:** Choose the **Tuned Adam MLP (relu, 128 hidden units)** for diagnosis.  


In [22]:
# Save Recall-First Tuned MLP Results
import joblib, pandas as pd, numpy as np

# Save tuned MLP model
model_filename = "recall_first_tuned_mlp.pkl"
joblib.dump(best_mlp, model_filename)

# Ensure 1D arrays for y_true and use predictions from the recall-first MLP
y_true = y_test.to_numpy() if hasattr(y_test, "to_numpy") else np.asarray(y_test)
y_pred_mlp = recall_first_y_pred
y_prob_mlp = recall_first_y_prob

# Optional gender column if present in test set
if isinstance(X_test, pd.DataFrame) and "gender" in X_test.columns:
    gender_vals = X_test["gender"].to_numpy()
else:
    gender_vals = np.full(shape=len(y_true), fill_value=np.nan)

# Build and save results DataFrame
results = pd.DataFrame({
    "gender": gender_vals,
    "y_true": y_true,
    "y_pred": y_pred_mlp,
    "y_prob": y_prob_mlp
})

preds_filename = "MendeleyData_50_50_MLP_recallfirst_predictions.csv"
results.to_csv(preds_filename, index=False)

print(f"Saved recall-first tuned MLP model → {model_filename}")
print(f"Saved predictions → {preds_filename}")

Saved recall-first tuned MLP model → recall_first_tuned_mlp.pkl
Saved predictions → MendeleyData_50_50_MLP_recallfirst_predictions.csv
